### PR TITLE
Cache Stage#summarize

### PR DESCRIPTION
**Pull request by wjordan**
Wraps Stage#summarize in a `Rails.cache.fetch` block. Includes `I18n.locale` in the cache key, since the result contains localized strings.

Local testing:

Before:

```
Started GET "/s/course2/stage/3/puzzle/2" for 127.0.0.1 at 2016-06-06 12:48:52 -0700
Processing by ScriptLevelsController#show as HTML
  Parameters: {"script_id"=>"course2", "stage_id"=>"3", "id"=>"2"}
  Rendered shared/_check_admin.html.haml (0.1ms)
  Rendered shared/_sharing_meta_tags.html.haml (0.1ms)
  Rendered levels/_reference_area.html.haml (0.4ms)
  Rendered levels/_apps_dependencies.html.haml (1.5ms)
  Rendered levels/_blockly.html.haml (10.1ms)
  Rendered levels/_admin.html.haml (1.4ms)
  Rendered levels/show.html.haml within layouts/application (16.5ms)
  Rendered layouts/_analytics.html.haml (0.6ms)
  Rendered layouts/_javascript_strings.haml (2.2ms)
  Rendered layouts/_environment_tag.html.haml (0.2ms)
  Rendered shared/_header_progress.html.haml (0.6ms)
  Rendered shared/_user_header_ajax.html.haml (0.1ms)
**Rendered layouts/_header.html.haml (115.9ms)
  Rendered layouts/_small_footer.html.haml (5.6ms)
  Rendered shared/_maybe_set_hoc_secret.html.haml (0.2ms)
Completed 200 OK in 189ms (Views: 165.3ms | ActiveRecord: 0.0ms)
```

After:

```
Started GET "/s/course2/stage/3/puzzle/2" for 127.0.0.1 at 2016-06-06 12:45:02 -0700
Processing by ScriptLevelsController#show as HTML
  Parameters: {"script_id"=>"course2", "stage_id"=>"3", "id"=>"2"}
  Rendered shared/_check_admin.html.haml (0.1ms)
  Rendered shared/_sharing_meta_tags.html.haml (0.1ms)
  Rendered levels/_reference_area.html.haml (0.3ms)
  Rendered levels/_apps_dependencies.html.haml (0.8ms)
  Rendered levels/_blockly.html.haml (6.2ms)
  Rendered levels/_admin.html.haml (0.6ms)
  Rendered levels/show.html.haml within layouts/application (10.3ms)
  Rendered layouts/_analytics.html.haml (1.1ms)
  Rendered layouts/_javascript_strings.haml (4.3ms)
  Rendered layouts/_environment_tag.html.haml (0.4ms)
  Rendered shared/_header_progress.html.haml (0.5ms)
  Rendered shared/_user_header_ajax.html.haml (0.1ms)
**Rendered layouts/_header.html.haml (6.0ms)
  Rendered layouts/_small_footer.html.haml (5.3ms)
  Rendered shared/_maybe_set_hoc_secret.html.haml (0.1ms)
Completed 200 OK in 59ms (Views: 49.2ms | ActiveRecord: 0.0ms)
```

This PR should greatly improve performance in a critical section of our server-side code - `ScriptLevelsController#show` is ~25% of our total Dashboard traffic response time, 50% of which is time spent in `header.html.haml`, ~95% of which is time spent in Stage#summarize. Based on this I would expect this optimization to boost the efficiency and response-time of our application servers by ~10% overall.

### CODE REVIEW

File: dashboard/app/models/stage.rb [hunk 1]
```diff
@@ -69,46 +69,49 @@ def lesson_plan_base_url
...
+          script_name: script.name,
+          script_stages: script.stages.to_a.size,
+          freeplay_links: script.freeplay_links,
+          id: id,
+          position: position,
+          name: localized_name,
+          title: localized_title,
+          flex_category: localized_category,
+          # Ensures we get the cached ScriptLevels, vs hitting the db
+          levels: script.script_levels.to_a.select{|sl| sl.stage_id == id}.map(&:summarize),
```
Comment (joshlory) [inline dashboard/app/models/stage.rb:84]: Are we going to end up double-caching ScriptLevels?

**Comment by joshlory** [comment]
Good find!  It looks like we'll be double-caching `ScriptLevel` (the object in the Script cache, and the summary in the Stage cache) but this seems better than the status quo.

If there isn't time to dig deeper I think it's ok to merge as-is.  We'll be trading space for perf which should be ok as long as there's space in the Rails cache.

**Comment by wjordan** [comment]
Thanks for the review.

I did some spot-profiling and couldn't find any single source of the slowdown (in particular no new ActiveRecord queries were generated), so my best guess is that the source of the slowdown was just a combination of the application logic and various localization-lookup calls plus Ruby garbage collection. (the localization is likely a bit slower since we're no longer using the optimized C-based i18n backend we had been using for a while.)

I'm not worried about the cache size - we have a limited number of levels/stages (under 1000), each of which will only be using a few kilobytes of memory for summary data. A quick test shows about 1MB total for summarized data across all stages (default memory cache-store size is 32MB with plenty of room to increase if needed). On production our level data is static with every app reload, so there shouldn't be any potential stale cache issues.

**Comment by joshlory** [comment]
Since the data is eventually sent to the client as JSON, could we push the content out to S3 (with the same fingerprints) instead of putting it in the Rails cache?

**Comment by wjordan** [comment]
If were were to invest more substantial refactoring time to this route in addition to this targeted one-line caching optimization, the first thing I would suggest (before pushing to S3) would be to take advantage of the `ApiController#script_structure` and `ApiController#user_progress` routes somewhere in `dashboard.header.build`, rather than calling `summarize_user_progress` and `script_level.stage.summarize` in the server-side page render.

Worth noting that the performance of `#script_structure` is also greatly improved by this caching PR (`1000ms` before, `20ms` after in my local testing course2 for example).

Beyond that, the `ApiController#script_structure` route could be replaced at some point by fetching an object from S3 directly. However, I'd expect the performance tradeoff between Rails-cached JSON and S3-hosted JSON to be negligible relative to other bottlenecks in our application (it could help during Hour of Code levels of scalability by reducing the total number of requests hitting our server, but I doubt it would impact typical end-user experience at all).

The downside to fetching stage content via separate Ajax call would be that it requires an extra client-side fetch after the `ScriptLevelsController#show` view, which would increase page-load time. However, this could be offset by caching the `#script_structure` data on the client and reusing across pages (I think we we currently do this for user-progress data already?).
